src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs

```diff
@@ -24,37 +21,69 @@ private DriverStateTable(ImmutableDictionary<object, IStateTable> tables)
+            internal GeneratorDriverState DriverState { get; }
+
+            public Compilation Compilation { get; }
+
+            public Builder(Compilation compilation, GeneratorDriverState driverState, ImmutableArray<ISyntaxInputNode> syntaxInputNodes, CancellationToken cancellationToken = default)
             {
-                _previousTable = previousTable;
+                Compilation = compilation;
+                DriverState = driverState;
+                _previousTable = driverState.StateTable;
+                _syntaxInputNodes = syntaxInputNodes;
                 _cancellationToken = cancellationToken;
             }
 
-            public void SetTable<T>(IIncrementalGeneratorNode<T> source, NodeStateTable<T> table)
+            public IStateTable GetSyntaxInputTable(ISyntaxInputNode syntaxInputNode)
             {
-                _tableBuilder[source] = table;
-            }
+                Debug.Assert(_syntaxInputNodes.Contains(syntaxInputNode));
 
-            public void AddInput<T>(InputNode<T> source, T value)
-            {
-                _tableBuilder[source] = source.CreateInputTable(_previousTable.GetStateTable(source), value);
-            }
+                // when we don't have a value for this node, we update all the syntax inputs at once
+                if (!_tableBuilder.ContainsKey(syntaxInputNode))
+                {
+                    // get a builder for each input node
+                    var builders = ArrayBuilder<ISyntaxInputBuilder>.GetInstance(_syntaxInputNodes.Length);
+                    foreach (var node in _syntaxInputNodes)
+                    {
+                        builders.Add(node.GetBuilder(_previousTable));
+                    }
+
+                    // update each tree for the builders, sharing the semantic model
+                    foreach ((var tree, var state) in GetLatestStateTableForNode(SharedInputNodes.SyntaxTrees))
+                    {
+                        var root = tree.GetRoot(_cancellationToken);
+                        var model = state != EntryState.Removed ? Compilation.GetSemanticModel(tree) : null;
+                        foreach (var builder in builders)
+                        {
+                            builder.VisitTree(root, state, model);
+                        }
+                    }
+
+                    // save the updated inputs
+                    foreach (var builder in builders)
+                    {
+                        builder.SaveStateAndFree(_tableBuilder);
+                    }
+                    builders.Free();
+                }
 
-            public void AddInput<T>(InputNode<T> source, IEnumerable<T> value)
-            {
-                _tableBuilder[source] = source.CreateInputTable(_previousTable.GetStateTable(source), value);
+                return _tableBuilder[syntaxInputNode];
             }
 
             public NodeStateTable<T> GetLatestStateTableForNode<T>(IIncrementalGeneratorNode<T> source)
```

Comment:
Should this method have a `Debug.Assert(source is not ISyntaxInputNode)`? Believe that those should always be going through `GetSyntaxInputTable` now.